### PR TITLE
feat: allow importing contract from address in studio UI

### LINKED ISSUE #616: Reset storage button confusion
Reset storage button in settings only restores the example contracts but not the ones made by the user. 
- Do we want the same functionality for the user generated intelligent contracts?
- Does it resets the browser cache?
- The button does not reset the database for listed contracts. Why not?

"All the examples will be restored" vs "following intelligent contracts will be removed". After pressing the button the contract files are still there.

Screenshot 2024-11-18 at 21 29 28

### PR DESCRIPTION
**Pull request by AgustinRamiroDiaz**
Watch the demo! https://www.loom.com/share/f3cdbbe61e2642af9f9ef6b1dbd95dbc?sid=7f453692-f9e5-4279-a7a0-4d44f8da2f4f

Fixes #616

# What

- Added ability to import deployed contracts from address in Studio UI
- Added `ImportContractModal` component for importing contracts via address
- Added `gen_getContractCode` RPC endpoint to retrieve deployed contract code
- Updated contract service and hooks to support imported contracts
- Fixed database handling to properly store and retrieve contract states as JSONB
- Improved database connection pooling and session management
- Added Web3 connection pooling for better resource management
- Fixed migration dependencies to avoid circular imports

# Why

- Users need to be able to import and interact with contracts deployed by others
- The UI was missing functionality to add existing contracts to the workspace
- Database connection handling needed improvements for reliability

# Testing done

- Tested importing contracts via the new import modal UI
- Verified imported contracts are properly displayed in contracts list
- Tested contract execution with imported contracts  
- Added integration tests for contract snapshot refresh functionality
- Updated transaction processor tests for improved error handling

# Decisions made

- Store contract code in base64 format extracted from state for deployed contracts
- Use JSONB format for account data storage instead of JSON strings
- Implement singleton Web3 connection pool to avoid connection exhaustion
- Added managed session context manager for proper database transaction handling

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with conventional commits

# Reviewing tips

- Focus on the new `ImportContractModal.vue` component for the main UI changes
- Review the `gen_getContractCode` endpoint implementation in `endpoints.py`
- Check the database session management improvements in `session_manager.py`
- The Web3 connection pooling in `web3_pool.py` addresses connection exhaustion issues

# User facing release notes

- **New Feature**: Import existing contracts by address - You can now add contracts deployed by others to your workspace by entering their address in the new import modal
- **Improvement**: Better database connection handling for improved stability
- **Fix**: Resolved issues with contract state storage and retrieval

## Summary by CodeRabbit

- **New Features**
  - Import deployed contracts by address (manual modal or via import-contract URL query); fetches on-chain code, creates/opens a contract file, records the deployment, and shows success/error notifications.
  - Programmatic import trigger from route query with automatic navigation/feedback.

- **UI/UX**
  - Import modal and toolbar button with real-time validation (invalid/duplicate cues), disabled/importing states, cancel/close resets, and inline feedback.

- **Chores**
  - Added a router helper dependency.

### CODE REVIEW

**Comment by kstroobants** [comment]
1. Can you check the rabbitai remarks? If they are resolved press on the button
2. Test edge case:
2.a. Input an address where the contract was not deployed so failed deployment transaction.
2.b. Input an address for a contract where you never sent a deployment transaction.
2.c. others...
3. Do not import contract if code does not exist

File: frontend/package.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
     "defu": "^6.1.4",
     "dexie": "^4.0.4",
     "floating-vue": "^5.2.2",
-    "genlayer-js": "0.15.1",
+    "genlayer-js": "0.16.0",
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/package.json]: it's exposed

File: frontend/src/components/contracts/ImportContractModal.vue [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    try {
+      // Try to fetch the actual contract code
+      contractCode = await fetchContractCode(contractAddress.value);
+      
+      notify({
+        title: 'Contract code retrieved',
+        text: 'Successfully fetched the contract code',
+        type: 'success',
+      });
+    } catch (codeError) {
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/components/contracts/ImportContractModal.vue]: addressed in fetchContractCode

File: frontend/src/hooks/useContractQueries.ts [hunk 3]
```diff
@@ -241,6 +241,20 @@ export function useContractQueries() {
...
+      if (!genlayerClient.value) {
+        throw new Error('Genlayer client not initialized');
+      }
+
+      return await genlayerClient.value.getContractCode(contractAddress as Address);
+    } catch (error) {
+      console.error('Error fetching contract code:', error);
+      throw error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/hooks/useContractQueries.ts:290]: It turns out it's not base64, I've tested it

File: frontend/src/services/ContractService.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      }
+
+      // Try to get contract schema to verify it exists
+      const schema = await client.getContractSchema(address as Address);
+      return !!schema;
+    } catch (error) {
+      console.error('Error verifying contract:', error);
+      return false;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/services/ContractService.ts]: fixed by removing the other one

File: frontend/src/services/ContractService.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+   * @returns Promise<boolean> indicating if contract exists
+   */
+  static async verifyContractExists(address: string): Promise<boolean> {
+    try {
+      const genlayer = useGenlayer();
+      const client = genlayer.client.value;
+      
+      if (!client) {
+        throw new Error('Genlayer client not initialized');
+      }
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/services/ContractService.ts]: fixed by removing the other one

File: frontend/src/services/ContractService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/services/ContractService.ts]: I don't understand this

File: frontend/src/components/contracts/ImportContractModal.vue [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        <input
+          id="contract-name"
+          v-model="contractName"
+          type="text"
+          placeholder="My Contract"
+          class="mt-1 block w-full rounded-md border-gray-300 shadow-sm focus:border-indigo-500 focus:ring-indigo-500 dark:border-zinc-600 dark:bg-zinc-800 dark:text-white sm:text-sm"
+        />
+        <p class="mt-1 text-sm text-gray-500">
+          Leave empty to use the address as the name
+        </p>
```
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/components/contracts/ImportContractModal.vue:112]: resolved

File: frontend/src/services/ContractService.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AgustinRamiroDiaz) [reply frontend/src/services/ContractService.ts]: removed entirely

**Comment by AgustinRamiroDiaz** [comment]
@kstroobants thanks for the review!


> Can you check the rabbitai remarks? If they are resolved press on the button

done!

> Test edge case:
> 2.a. Input an address where the contract was not deployed so failed deployment transaction.

done! works as expected (you get an error that you cannot retrieve the contract code)

> 2.b. Input an address for a contract where you never sent a deployment transaction.

done! works as expected (you get an error that you cannot retrieve the contract code)

> 2.c. others...

tested also with: 
- happy path
- setting a wrong address (the UI doesn't even let you)

> Do not import contract if code does not exist

Done!

**Comment by kstroobants** [comment]
Unused imports:
- ContractService is imported but never used
- props is defined but not used

**Comment by AgustinRamiroDiaz** [comment]
@kstroobants done! 

>  Unused imports:
> ContractService is imported but never used
> props is defined but not used

Why wasn't this caught by the linter?

**Comment by AgustinRamiroDiaz** [comment]
@kstroobants e2e tests have passed, but not the new "Load Tests"
Can we go ahead and merge this?

**Comment by kstroobants** [comment]
> @kstroobants e2e tests have passed, but not the new "Load Tests" Can we go ahead and merge this?

@cristiam86 I made this PR https://github.com/genlayerlabs/genlayer-studio/pull/1322 for the load tests.

**Comment by cristiam86** [comment]
Yes, we are just about to merge a PR that migrates from flask to fastapi. After that one, we can merge this one. 
Don't worry, I'm also very interested :)

**Review by cristiam86 — APPROVED** [review]
